hatch/models.py

```diff
@@ -0,0 +1,62 @@
+import hashlib
+from typing import List
+
+from pydantic import BaseModel
+
+from hatch import config
+
+
+def get_sha(path):
+    """Calculate and cache the sha256 hash of a file.
+
+    We cache because a big workspace can have lots of large files, and serving
+    the index would involve calculated the hashes of all files, which could get
+    slow.
+
+    We cache it in a separate directory, but the same relative path. So the
+    cache for config.WORKSPACES/workspace1/output/file.txt is located at
+    config.CACHE/workspace1/output/file.txt.
+    """
+    sha_path = config.CACHE / path.relative_to(config.WORKSPACES)
+    # does cache file exist and is current?
+    if sha_path.exists() and sha_path.stat().st_mtime > path.stat().st_mtime:
```

Comment:
Could you break this out into a variable? I don't know the stat or st_mtime bits particularly well.  I think it's checking the cache file was modified more recently than the workspace file?  Would be good to make that more explicit though 🙂

Reply:
done